### PR TITLE
Allow move toolstrips (take 2)

### LINKED ISSUE #8680: Toolbar position is not saved or restored
## Current behaviour

Since #8572, I [always](https://github.com/gitextensions/gitextensions/pull/8572#issuecomment-734340761) end up with this toolbar at startup:
image

Even if I move the toolbars. Even when only on instance is opened.

Even with #8679 fix (to only save the toolbar settings)

## Expected behaviour

I can configure the toolbar position and it is saved (or restored)  well.

## Steps to reproduce

1. Get a dev version of GE (portable from CI or your custom version) and launch it
2. Change toolbar setup
3. Restart
4. See toolbar position configuration is the same as initally

## Screenshots

See above.

## Did this work in previous version of GitExtensions

Never since the feature was introduced... 

## Environment

- Git Extensions 33.33.33
- Build 3245408651bd658ee636e7c3b57f3a02bbe4abf4
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.8.4261.0
- DPI 192dpi (200% scaling)

## Diagnostics

The main goal to opening this issue is:

How can I debug it easily?
Where the configuration data are stored?

### PR DESCRIPTION
**Pull request by RussKie**
Reintroduce ability to move toolstrips (as proposed in #8572).

This implementation is heavily influenced by the original implementation [`ToolStripSettingsManager`](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/ToolStripSettingsManager.cs) in Windows Forms, however we only track visibility and location of each individual toolstrip. We're also only saving specified toolstrips as opposed to saving all known toolstrips (which in various situations results in terminal failures to restore toolstrips information).

There are two key changes:

1. Restore toolstrip locations after the form's geometry has been restored.    This ensures the toolstrips can be placed at the same coords as they    were saved.
2. Hosting ToolStripPanel's padding must be set to 0 before toolstrips'    locations can be restored/saved. This mitigates https://github.com/dotnet/winforms/issues/4449

Fixes #8680
Reverts #8719

## Test methodology 

- 
- 
- 

## Test environment(s) 

- GIT 
- Windows 

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by gerhardol** [comment]
No toolbar at all for me.
Asserts on GitUI\ToolStripExSettingsManager.cs:46 at startup after the first
Not investigated, toolStripSettingsToApply.Count==3

No quick opinion after browsing the changes

**Comment by RussKie** [comment]
> No toolbar at all for me.
> Asserts on GitUI\ToolStripExSettingsManager.cs:46 at startup after the first
> Not investigated, toolStripSettingsToApply.Count==3

Should be fixed now. If you could try again - it'd be great.
I also provided an ability to rest the toolstrip layout in case it gets messy. The layout is also reset whenever the config is deemed to be invalid.
image

Toolstrips are extremely finicky. There are so many different caveats and edge cases... I'm starting to understand why this UX paradigm was left behind.
I _think_ I got the major scenarios ironed out - i.e. it is possible to move toolstrips around and their positions are persisted. There's the first (of many) caveat - if there are several toolstrips on the same row, and there's a gap between the toolstrips - the right most toolstrip will be shifted by an arbitrary amount of pixels when the layout is restored. From what I understand this is caused by the fact that the toolstrips have variable length - e.g., the repo and branch selectors are of variable length, the commit button and stash button can be of variable size, etc. All of these variable lengths can amount to hundreds of pixels.
I don't personally consider a layout with gaps to be major use case, and my care factor starts rapidly descend to zero after so many frustrating hours spent dealing with this.

**Comment by RussKie** [comment]
I'm leaning over to putting this feature behind a feature flag, and introducing a "Preview" or "Experimental" settings (and a respective tab) where this feature can be opted-in.

**Review by gerhardol — COMMENTED** [review]
+1
Reset is not working for me.
Some usage and code review, not really any comments.

> I'm leaning over to putting this feature behind a feature flag, and introducing a "Preview" or "Experimental" settings (and a respective tab) where this feature can be opted-in.

If it seems to work OK, it can be official for me. However, I do not find this to be a very important feature, all buttons fit by default.

**Review by mstv — COMMENTED** [review]
+1 Works well for my simple use case (default layout).
(If the toolbars are stacked vertically, the may be restored in a different sequence.)

> Reset is not working for me.

Works well for me. Is your "command toolstrip" empty, Gerhard?

> If it seems to work OK, it can be official for me.

If it works for your use case, Igor, I think it is good enough.

**Comment by pmiossec** [comment]
Some strange behaviors:

Prerequisite:  No toolbars hidden

1. At each restart the toolbars are moved on the right. After 3 restarts:

image


Prerequisite:  "Filter" toolbar hidden


2. On 2nd restart (and after):

image

3. Clicking on "Reset" always end up with:

image

4. AFter a "Reset", on the next start, toolbars are inverted:

image

5. And after another restart:

image

6. After another one:

image

So, to conclude, with these changes, toolbars seems to be fed up with the lockdown 😉 

Reminder on my configuration (maybe due to the scaling):

- Git Extensions 33.33.33
- Build ecad3e66451d2269f54fc1f4066b829d96771e4b
- Git 2.34.0.windows.1
- Microsoft Windows NT 10.0.19042.0
- .NET 5.0.13
- DPI 192dpi (200% scaling)

File: GitUI/Theming/AppColorExtension.cs [hunk 1]
```diff
@@ -1,11 +1,19 @@
...
-            ThemeModule.Settings.Theme.GetColor(name);
+        public static Color GetThemeColor(this AppColor name)
+        {
+            if (AppSettings.IsDesignMode)
+            {
+                return Color.Transparent;
+            }
+
+            return ThemeModule.Settings.Theme.GetColor(name);
+        }
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/Theming/AppColorExtension.cs:17]: It is related in a sense that it allows opening the main form in the designer.

File: GitUI/TranslatedStrings.cs [hunk 2]
```diff
@@ -98,6 +98,8 @@ internal sealed class TranslatedStrings : Translate
         private readonly TranslationString _resetSelectedLinesConfirmation = new("Are you sure you want to reset the changes to the selected lines?");
         private readonly TranslationString _resetChangesCaption = new("Reset changes");
 
+        private readonly TranslationString _resetToolbarsCaption = new("Reset");
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/TranslatedStrings.cs:101]: More like "Reset toolbars (toolstrips?) to the original layout". I don't really want to emphasize positions here.

**Comment by RussKie** [comment]
Thank you @pmiossec! I really appreciate your feedback. These behaviours are bizarre to say the least. I indeed haven't tested on non standard dpi, and the scale factor may be causing havoc.
Back to the drawing board.... ugh... 😒

**Pull request closed without merging**